packages/server/utils/RedisLockQueue.ts

```diff
@@ -0,0 +1,107 @@
+import sleep from '../../client/utils/sleep'
+import generateUID from '../generateUID'
+import getRedis from './getRedis'
+
+export default class RedisLockQueue {
+  uid = generateUID()
+  uidWithTimestamp: string | undefined
+  queueKey: string
+  waitTimeMs: number
+  ttl: number
+  queued = false
+  timeoutId: NodeJS.Timeout | undefined
+  uidToFlush = ''
+  redis = getRedis()
+
+  // Check lock every 100 ms
+  private static readonly checkAfterMs = 100
+  private static readonly timestampDelimiter = '::'
+
+  constructor(key: string, ttl: number) {
+    this.queueKey = `lock:queue:${key}`
+    this.ttl = ttl
+    this.waitTimeMs = 0
+  }
+
+  unlock = async () => {
+    if (this.timeoutId) {
+      clearTimeout(this.timeoutId)
+    }
+    // Remove all occurrences of the event from the list
+    return this.redis.lrem(this.queueKey, 0, this.uidWithTimestamp ?? this.uid)
+  }
+
+  private markTaskAsRunning = async () => {
+    // Append a timestamp to uid and replace the head
+    this.uidWithTimestamp = `${this.uid}::${Date.now()}`
+    return this.redis.lset(this.queueKey, 0, this.uidWithTimestamp)
+  }
+
+  // Removes the head
+  private flushHead = async () => {
+    return this.redis.lpop(this.queueKey)
+  }
+
+  private setLockAndValidateQueue = async () => {
+    // Add an event to the queue only once
+    if (!this.queued) {
+      const resultingListLength = await this.redis.rpush(this.queueKey, this.uid)
+      this.queued = true
+      // If the resulting LIST length is 1, we're already first on the list, return no lock
+      return resultingListLength !== 1
+    }
+    // Checking if we are the first on the list
+    const head = await this.redis.lindex(this.queueKey, 0)
+    if (head == null) {
+      // If no head element exists, the queue was removed for some reason
+      // Now we cannot guarantee the correct order of the event
+      throw new Error(`Could not achieve lock for ${this.queueKey}. Queue does not exists`)
+    }
+
+    const headUid = head.split(RedisLockQueue.timestampDelimiter)[0]
+    if (headUid === this.uid) {
```

Comment:
let's do `head === this.uid` because it tells us that no timestamp has been added yet